drivers/storage/azure/storage/azure_storage.go

```diff
@@ -0,0 +1,914 @@
+	}
+	if result := d.config.GetString(azure.ConfigAZUREContainerKey); result != "" {
+		return result
+	}
+	return "vhds"
+}
+
+func (d *driver) getClientID() string {
+	if result := os.Getenv("AZURE_CLIENT_ID"); result != "" {
+		return result
+	}
+	return d.config.GetString(azure.ConfigAZUREClientIDKey)
+}
+
+func (d *driver) getClientSecret() string {
+	if result := os.Getenv("AZURE_CLIENT_SECRET"); result != "" {
+		return result
+	}
+	return d.config.GetString(azure.ConfigAZUREClientSecretKey)
+}
+
+func (d *driver) getCertPath() string {
+	if result := os.Getenv("AZURE_CERT_PATH"); result != "" {
+		return result
+	}
+	return d.config.GetString(azure.ConfigAZURECertPathKey)
+}
+
+func (d *driver) getUseHTTPS() bool {
+	if result := os.Getenv("AZURE_USE_HTTPS"); result != "" {
+		return result == "true"
+	}
+	return d.config.GetBool(azure.ConfigAZUREUseHTTPSKey)
+}
+
+func (d *driver) tag() string {
+	return d.config.GetString(azure.ConfigAZURETagKey)
+}
+
+// TODO rexrayTag
+/*func (d *driver) rexrayTag() string {
+  return d.config.GetString("azure.rexrayTag")
+}*/
+
+var errGetLocDevs = goof.New("error getting local devices from context")
+
+func (d *driver) toTypesVolume(
+	ctx types.Context,
+	blobs *[]blobStorage.Blob,
+	attachments types.VolumeAttachmentsTypes) ([]*types.Volume, error) {
+
+	/* TODO:
+	var (
+		ld *types.LocalDevices
+		ldOK bool
+	)
+
+	if attachments.Devices() {
+		// Get local devices map from context
+		if ld, ldOK = context.LocalDevices(ctx); !ldOK {
+			return nil, errGetLocDevs
+		}
+	}
+	*/
+
+	var volumesSD []*types.Volume
+	for _, blob := range *blobs {
+		volumeSD, error := d.toTypeVolume(ctx, &blob, attachments)
+		if error != nil {
+			ctx.WithError(error).Error("Failed to convert volume")
+		} else if volumeSD != nil {
+			volumesSD = append(volumesSD, volumeSD)
+		}
+	}
+	return volumesSD, nil
+}
+
+func (d *driver) toTypeVolume(
+	ctx types.Context,
+	blob *blobStorage.Blob,
+	attachments types.VolumeAttachmentsTypes) (*types.Volume, error) {
+
+	// Metadata can have these fileds:
+	// microsoftazurecompute_resourcegroupname:trex
+	// microsoftazurecompute_vmname:ttt
+	// microsoftazurecompute_disktype:DataDisk (or OSDisk)
+	// microsoftazurecompute_diskid:7d9df1c9-7b4f-41d4-a2e3-6870dfa573ba
+	// microsoftazurecompute_diskname:ttt-20161221-130722
+	// microsoftazurecompute_disksizeingb:50
+
+	btype := blob.Metadata["microsoftazurecompute_disktype"]
+	if btype == "" && !strings.HasSuffix(blob.Name, vhdExtension) {
+		return nil, nil
+	}
+	attachState := types.VolumeAvailable
+	bstate := "detached"
+	if blob.Metadata["microsoftazurecompute_vmname"] != "" {
+		bstate = "attached"
+		attachState = types.VolumeAttached
+	}
+	var attachmentsSD []*types.VolumeAttachment
+	if attachments.Requested() {
+		if attachState == types.VolumeAttached {
+			attachedInstanceID := types.InstanceID{
+				ID:     blob.Metadata["microsoftazurecompute_vmname"],
+				Driver: d.name,
+			}
+			if attachments.Mine() {
```

Comment:
Is it necessary for the Azure driver to do this filtering itself? I don't think that it is, and I don't think that it needs to handle setting `attachState` either. The framework handles this for you, here: https://github.com/codedellemc/libstorage/blob/release/0.4.0/api/server/router/volume/volume_routes.go#L131-L209

If a driver really does need to calculate `attachState` itself, it can do so. That will override the framework's default behavior of saying the volume is available if there are no attachments returned, it is attached if it is attached to the IID of the caller, or unavailable if attached to a different IID.

It is the intended behavior that `attachState` is not set if `attachments.Requested()` is false.